prov/gni/src/gnix_sep.c

```diff
@@ -0,0 +1,860 @@
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*
+ * Endpoint common code
+ */
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#include "gnix.h"
+#include "gnix_cm_nic.h"
+#include "gnix_ep.h"
+#include "gnix_vc.h"
+#include "gnix_util.h"
+#include "gnix_msg.h"
+
+/******************************************************************************
+ * Forward declaration for ops structures.
+ ******************************************************************************/
+
+static struct fi_ops gnix_sep_fi_ops;
+static struct fi_ops_ep gnix_sep_ops;
+/*
+static struct fi_ops gnix_tx_fi_ops;
+static struct fi_ops_ep gnix_tx_ops;
+*/
+static struct fi_ops_cm gnix_sep_rxtx_cm_ops;
+static struct fi_ops_msg gnix_sep_msg_ops;
+static struct fi_ops_rma gnix_sep_rma_ops;
+static struct fi_ops_tagged gnix_sep_tagged_ops;
+static struct fi_ops_atomic gnix_sep_atomic_ops;
+
+/*******************************************************************************
+ * SEP(EP) OPS API function implementations.
+ ******************************************************************************/
+/* TODO:
+	initialize capabilities for tx_priv?
+	initialize attr?
+*/
+
+static void __trx_destruct(void *obj)
+{
+	int __attribute__((unused)) ret;
+	struct gnix_fid_trx *trx = (struct gnix_fid_trx *) obj;
+	struct gnix_fid_ep *ep_priv;
+	struct gnix_fid_sep *sep_priv;
+	struct fid_domain *domain;
+	int refs_held;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	ep_priv = trx->ep;
+	assert(ep_priv != NULL);
+	sep_priv = trx->sep;
+	assert(sep_priv != NULL);
+	domain = sep_priv->domain;
+	assert(domain != NULL);
+
+	refs_held = _gnix_ref_put(ep_priv);
+	if (refs_held == 0)
+		_gnix_ref_put(sep_priv->cm_nic);
+	_gnix_ref_put(sep_priv);
+
+	free(trx);
+}
+
+static int gnix_sep_tx_ctx(struct fid_ep *sep, int index,
+			   struct fi_tx_attr *attr,
+			   struct fid_ep **tx_ep, void *context)
+{
+	int ret = FI_SUCCESS;
+	struct gnix_fid_sep *sep_priv;
+	struct gnix_fid_ep *ep_priv = NULL;
+	struct gnix_fid_trx *tx_priv = NULL;
+	struct fid_ep *ep_ptr;
+	struct gnix_ep_attr ep_attr = {0};
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	sep_priv = container_of(sep, struct gnix_fid_sep, ep_fid);
+	if ((sep_priv->ep_fid.fid.fclass != FI_CLASS_SEP) ||
```

Comment:
It is passed from the user so it wouldn't hurt to check it. The user would have to have forgotten to call fi_scalable_ep first.